Build failure: python3Packages.requests for < Python3.12 on Darwin (ModuleNotFoundError: No module named '_ctypes')

### Nixpkgs version

- Unstable (25.05)

### Describe the bug

I'm on aarch64-darwin and one of my packages started to fail to build after I did this flake.lock update
```
• Updated input 'nixpkgs':
    'github:nixos/nixpkgs/e3e32b642a31e6714ec1b712de8c91a3352ce7e1?narHash=sha256-7rTAMNTY1xoBwz0h7ZMtEcd8LELk9R5TzBPoHuhNSCk%3D' (2025-03-09)
  → 'github:nixos/nixpkgs/c80f6a7e10b39afcc1894e02ef785b1ad0b0d7e5?narHash=sha256-C7jVfohcGzdZRF6DO%2BybyG/sqpo1h6bZi9T56sxLy%2Bk%3D' (2025-03-15)
```

error is
```
ModuleNotFoundError: No module named '_ctypes'
```
Package I use requires Python 3.10, however issue is reproducable on Python 3.11 as well. Python 3.12 is not affected. Below I'm including reproducer for Python 3.11, as it's simpler to do it there.

### Steps to reproduce

```
❯ nix shell nixpkgs#python311Packages.requests
error: builder for '/nix/store/xq9g48g7383zpqcvfp39925yvq4zaapl-python3.11-cffi-1.17.1.drv' failed with exit code 1;
       last 25 log lines:
       >     return run_commands(dist)
       >            ^^^^^^^^^^^^^^^^^^
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/_distutils/core.py", line 202, in run_commands
       >     dist.run_commands()
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 983, in run_commands
       >     self.run_command(cmd)
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/dist.py", line 999, in run_command
       >     super().run_command(command)
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/_distutils/dist.py", line 1002, in run_command
       >     cmd_obj.run()
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/command/bdist_wheel.py", line 416, in run
       >     impl_tag, abi_tag, plat_tag = self.get_tag()
       >                                   ^^^^^^^^^^^^^^
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/command/bdist_wheel.py", line 330, in get_tag
       >     plat_name = get_platform(self.bdist_dir)
       >                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
       >   File "/nix/store/n7wlvi3n7551v6ngy68vwrya2l8q0nmz-python3.11-setuptools-75.8.0/lib/python3.11/site-packages/setuptools/command/bdist_wheel.py", line 68, in get_platform
       >     from wheel.macosx_libfile import calculate_macosx_platform_tag
       >   File "/nix/store/m2ap50pixr7kvcmiabsd2rp9nzn62ryl-python3.11-wheel-0.45.1/lib/python3.11/site-packages/wheel/macosx_libfile.py", line 43, in <module>
       >     import ctypes
       >   File "/nix/store/h7hmj8v6rgddxknk3cyxw0wl2lzwh9wl-python3-3.11.11/lib/python3.11/ctypes/__init__.py", line 8, in <module>
       >     from _ctypes import Union, Structure, Array
       > ModuleNotFoundError: No module named '_ctypes'
       >
       > ERROR Backend subprocess exited when trying to invoke build_wheel
       For full logs, run:
         nix-store -l /nix/store/xq9g48g7383zpqcvfp39925yvq4zaapl-python3.11-cffi-1.17.1.drv
error: 1 dependencies of derivation '/nix/store/84cqa4w1rxfvn4w3jc3ia1a4r37xhwd7-python3.11-brotlicffi-1.1.0.0.drv' failed to build
error: 1 dependencies of derivation '/nix/store/mmn8m3klsgk2fjvqwx43nxwf5535wyb0-python3.11-gevent-24.11.1.drv' failed to build
error: 1 dependencies of derivation '/nix/store/i9bdrl5g1s3a5ixkim8an5jlyg2aqkhr-python3.11-requests-2.32.3.drv' failed to build
```

### Expected behaviour

Build success

### Screenshots

_No response_

### Relevant log output

```console

```

### Additional context

_No response_

### System metadata

```
 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.3.0, macOS 15.3.2`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.26.3`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/store/ilwpca5ykjz25f0sd55g66k25gpmi3ld-nixpkgs/nixpkgs`
```

### Are you using nix-darwin?

No, I am not using nix-darwin.

### Notify maintainers

@fab

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+bug%22+label%3A%226.topic%3A+darwin%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

Failure when building python:

```
clang -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/include -fno-semantic-interposition -flto -std=c11 -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wstrict-prototypes -Werror=implicit-function-declarat
ion -fvisibility=hidden -fno-semantic-interposition -I./Include/internal -I_ctypes/darwin -I/nix/store/0bpf12qridbmw2h8cdw82rqh9p1mfm6i-libffi-39-dev/include -I./Include -I. -I/nix/store/c1glyynad4rsdib6690zxhhb6jwmcl1z-bzip2-1.0.8-dev/include -I/nix/store/g7134ihxfgc1pi7ygs63ck2l3sms0mrv-expat-2.6.
4-dev/include -I/nix/store/0bpf12qridbmw2h8cdw82rqh9p1mfm6i-libffi-39-dev/include -I/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/include -I/nix/store/m2kwjq423w2hsbkbmj2y2080k99ks5fk-mpdecimal-4.0.0-dev/include -I/nix/store/r55vh3zhcsr94jvqxf1vfk8cf4r161q4-ncurses-6.5-dev/include -I/
nix/store/p5kl6l6yx56m1p3ngnrahb3s6sq2qqmj-openssl-3.4.1-dev/include -I/nix/store/7659jmimxp2zcbg0g513hk8g0zpxwaha-sqlite-3.48.0-dev/include -I/nix/store/148yzmvyi2i9n182p7d20x14x1lmry03-xz-5.6.4-dev/include -I/nix/store/r1nhxwfzgyipxz74mkv153hmmy8f7hvg-zlib-1.3.1-dev/include -I/nix/store/mrnbn0z23c
sii8myzvknp0ci2a1z65i4-tzdata-2025a-dev/include -I/nix/store/zw8vp9w213ci8ggnj6faqbixrjaf5w1s-gdbm-1.24-dev/include -I/nix/store/69x34bsdvnrsw2p4sqyknqr00nn41sjz-readline-8.2p13-dev/include -I/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Include -I/private/tmp/nix-shell.Ik1vjx/tmp.CpID
pkp7fC/Python-3.11.11 -c /private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_ctypes/callbacks.c -o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_ctypes/callbacks.o -DUSING_MALLOC_CLOSURE_DOT_C=1 -DMACOSX -DHAVE_FFI_PREP_CIF_VAR
=1 -DHAVE_FFI_PREP_CLOSURE_LOC=1 -DHAVE_FFI_CLOSURE_ALLOC=1
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'
clang -bundle -undefined dynamic_lookup -L/nix/store/5988q68ifr0plfhb5xm2vakhkhjvj6jd-bzip2-1.0.8/lib -L/nix/store/zx5lwbqh7bkkr8kyg1sia5bx8v3419mi-expat-2.6.4/lib -L/nix/store/b8cvkxggxhzmy9vwabswn189cvbz05b7-libffi-39/lib -L/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/lib -L/nix/st
ore/3gnsdhgdwkl49dx8xd4bw3fzayswirxs-mpdecimal-4.0.0/lib -L/nix/store/011j6gxha0k91jqfqzv36zl4hlb0gp2l-ncurses-6.5/lib -L/nix/store/632hkjxqm9877a1l1fvbjw00hp5z9srw-openssl-3.4.1/lib -L/nix/store/7wm2il6yjbnnj9x3wmqnvkwg9p0fqh8p-sqlite-3.48.0/lib -L/nix/store/5lmmpz34b432kp6qssk11cmnxch93syj-xz-5.6.
4/lib -L/nix/store/rssgki46hkwsk62j4id06dr2byc44nvf-zlib-1.3.1/lib -L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib -L/nix/store/ijc51vj231rf5psd2k75rjlci5bj7af8-gdbm-1.24-lib/lib -L/nix/store/fbnwbv2jnxy1nfwvs1fk1iq3d15nv4dm-readline-8.2p13/lib -L/nix/store/5988q68ifr0plfhb5xm2vakhkhj
vj6jd-bzip2-1.0.8/lib -L/nix/store/zx5lwbqh7bkkr8kyg1sia5bx8v3419mi-expat-2.6.4/lib -L/nix/store/b8cvkxggxhzmy9vwabswn189cvbz05b7-libffi-39/lib -L/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/lib -L/nix/store/3gnsdhgdwkl49dx8xd4bw3fzayswirxs-mpdecimal-4.0.0/lib -L/nix/store/011j6gxha0
k91jqfqzv36zl4hlb0gp2l-ncurses-6.5/lib -L/nix/store/632hkjxqm9877a1l1fvbjw00hp5z9srw-openssl-3.4.1/lib -L/nix/store/7wm2il6yjbnnj9x3wmqnvkwg9p0fqh8p-sqlite-3.48.0/lib -L/nix/store/5lmmpz34b432kp6qssk11cmnxch93syj-xz-5.6.4/lib -L/nix/store/rssgki46hkwsk62j4id06dr2byc44nvf-zlib-1.3.1/lib -L/nix/store/
k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib -L/nix/store/ijc51vj231rf5psd2k75rjlci5bj7af8-gdbm-1.24-lib/lib -L/nix/store/fbnwbv2jnxy1nfwvs1fk1iq3d15nv4dm-readline-8.2p13/lib -flto -Wl,-export_dynamic -g -L/nix/store/5988q68ifr0plfhb5xm2vakhkhjvj6jd-bzip2-1.0.8/lib -L/nix/store/zx5lwbqh7bkkr8ky
g1sia5bx8v3419mi-expat-2.6.4/lib -L/nix/store/b8cvkxggxhzmy9vwabswn189cvbz05b7-libffi-39/lib -L/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/lib -L/nix/store/3gnsdhgdwkl49dx8xd4bw3fzayswirxs-mpdecimal-4.0.0/lib -L/nix/store/011j6gxha0k91jqfqzv36zl4hlb0gp2l-ncurses-6.5/lib -L/nix/store
/632hkjxqm9877a1l1fvbjw00hp5z9srw-openssl-3.4.1/lib -L/nix/store/7wm2il6yjbnnj9x3wmqnvkwg9p0fqh8p-sqlite-3.48.0/lib -L/nix/store/5lmmpz34b432kp6qssk11cmnxch93syj-xz-5.6.4/lib -L/nix/store/rssgki46hkwsk62j4id06dr2byc44nvf-zlib-1.3.1/lib -L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib -
L/nix/store/ijc51vj231rf5psd2k75rjlci5bj7af8-gdbm-1.24-lib/lib -L/nix/store/fbnwbv2jnxy1nfwvs1fk1iq3d15nv4dm-readline-8.2p13/lib -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -I/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/include -I/nix/store/c1glyynad4rsdib6690zxhh
b6jwmcl1z-bzip2-1.0.8-dev/include -I/nix/store/g7134ihxfgc1pi7ygs63ck2l3sms0mrv-expat-2.6.4-dev/include -I/nix/store/0bpf12qridbmw2h8cdw82rqh9p1mfm6i-libffi-39-dev/include -I/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/include -I/nix/store/m2kwjq423w2hsbkbmj2y2080k99ks5fk-mpdecimal-4
.0.0-dev/include -I/nix/store/r55vh3zhcsr94jvqxf1vfk8cf4r161q4-ncurses-6.5-dev/include -I/nix/store/p5kl6l6yx56m1p3ngnrahb3s6sq2qqmj-openssl-3.4.1-dev/include -I/nix/store/7659jmimxp2zcbg0g513hk8g0zpxwaha-sqlite-3.48.0-dev/include -I/nix/store/148yzmvyi2i9n182p7d20x14x1lmry03-xz-5.6.4-dev/include -I
/nix/store/r1nhxwfzgyipxz74mkv153hmmy8f7hvg-zlib-1.3.1-dev/include -I/nix/store/mrnbn0z23csii8myzvknp0ci2a1z65i4-tzdata-2025a-dev/include -I/nix/store/zw8vp9w213ci8ggnj6faqbixrjaf5w1s-gdbm-1.24-dev/include -I/nix/store/69x34bsdvnrsw2p4sqyknqr00nn41sjz-readline-8.2p13-dev/include build/temp.macosx-11
.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/blob.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/connection.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Pyth
on-3.11.11/Modules/_sqlite/cursor.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/microprotocols.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/module.o build/temp.macosx-11
.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/prepare_protocol.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/row.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC
/Python-3.11.11/Modules/_sqlite/statement.o build/temp.macosx-11.3-arm64-3.11/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_sqlite/util.o -L. -L/Users/ihrachys/src/nixpkgs/outputs/out/lib -L/nix/store/5988q68ifr0plfhb5xm2vakhkhjvj6jd-bzip2-1.0.8/lib -L/nix/store/zx5lwbqh7bkkr8k
yg1sia5bx8v3419mi-expat-2.6.4/lib -L/nix/store/b8cvkxggxhzmy9vwabswn189cvbz05b7-libffi-39/lib -L/nix/store/ma585pmsp3iwdiwazgig20kj1nsb37ss-libxcrypt-4.4.38/lib -L/nix/store/3gnsdhgdwkl49dx8xd4bw3fzayswirxs-mpdecimal-4.0.0/lib -L/nix/store/011j6gxha0k91jqfqzv36zl4hlb0gp2l-ncurses-6.5/lib -L/nix/stor
e/632hkjxqm9877a1l1fvbjw00hp5z9srw-openssl-3.4.1/lib -L/nix/store/7wm2il6yjbnnj9x3wmqnvkwg9p0fqh8p-sqlite-3.48.0/lib -L/nix/store/5lmmpz34b432kp6qssk11cmnxch93syj-xz-5.6.4/lib -L/nix/store/rssgki46hkwsk62j4id06dr2byc44nvf-zlib-1.3.1/lib -L/nix/store/ijc51vj231rf5psd2k75rjlci5bj7af8-gdbm-1.24-lib/lib
 -L/nix/store/fbnwbv2jnxy1nfwvs1fk1iq3d15nv4dm-readline-8.2p13/lib -lsqlite3 -o build/lib.macosx-11.3-arm64-3.11/_sqlite3.cpython-311-darwin.so
clang: warning: argument unused during compilation: '-fno-semantic-interposition' [-Wunused-command-line-argument]
clang: warning: argument unused during compilation: '-fno-semantic-interposition' [-Wunused-command-line-argument]
/private/tmp/nix-shell.Ik1vjx/tmp.CpIDpkp7fC/Python-3.11.11/Modules/_ctypes/callbacks.c:438:18: error: call to undeclared function 'ffi_prep_closure'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  438 |         result = ffi_prep_closure(p->pcl_write, &p->cif, closure_fcn, p);
      |                  ^
1 error generated.
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'
ld: warning: directory not found for option '-L/nix/store/k6dpbrxznfk6yq3xk7i6lhfpvfxg9lvj-tzdata-2025a/lib'

The necessary bits to build these optional modules were not found:
_tkinter                                                       
To find the necessary bits, look in setup.py in detect_modules() for the module's name.


Failed to build these modules:
_ctypes                                                        
```

(the build still succeeds)

--- a/pkgs/development/interpreters/python/cpython/default.nix
+++ b/pkgs/development/interpreters/python/cpython/default.nix
@@ -434,6 +434,9 @@ in with passthru; stdenv.mkDerivation (finalAttrs: {
     for path in /usr /sw /opt /pkg; do
       substituteInPlace ./setup.py --replace-warn $path /no-such-path
     done
+  '' + optionalString (stdenv.hostPlatform.isDarwin && pythonOlder "3.12") ''
+    # Fix _ctypes module compilation
+    export NIX_CFLAGS_COMPILE+=" -DUSING_APPLE_OS_LIBFFI=1"
   '' + optionalString stdenv.hostPlatform.isDarwin ''
     # Override the auto-detection in setup.py, which assumes a universal build
     export PYTHON_DECIMAL_WITH_MACHINE=${if stdenv.hostPlatform.isAarch64 then "uint128" else "x64"}

